src/test/scala/gitbucket/core/api/ApiSpecModels.scala

```diff
@@ -42,228 +48,609 @@ object ApiSpecModels {
+       |"path":"README.md",
+       |"commit_id":"6dcb09b5b57875f334f61aebed695e2e4193db5e",
+       |"user":$jsonUser,
+       |"body":"Maybe you should use more emoji on this line.",
+       |"created_at":"2015-05-05T23:40:27Z",
+       |"updated_at":"2015-05-05T23:40:27Z",
+       |"url":"http://gitbucket.exmple.com/api/v3/repos/octocat/Hello-World/pulls/comments/29724692",
+       |"html_url":"http://gitbucket.exmple.com/octocat/Hello-World/pull/1347#discussion_r29724692",
+       |"pull_request_url":"http://gitbucket.exmple.com/api/v3/repos/octocat/Hello-World/pulls/1347",
+       |"_links":{
+         |"self":{"href":"http://gitbucket.exmple.com/api/v3/repos/octocat/Hello-World/pulls/comments/29724692"},
+         |"html":{"href":"http://gitbucket.exmple.com/octocat/Hello-World/pull/1347#discussion_r29724692"},
+         |"pull_request":{"href":"http://gitbucket.exmple.com/api/v3/repos/octocat/Hello-World/pulls/1347"}}
+       |}""".stripMargin
+
+  val jsonComment = s"""{
+       |"id":1,
+       |"user":$jsonUser,
+       |"body":"Me too",
+       |"created_at":"2011-04-14T16:00:49Z",
+       |"updated_at":"2011-04-14T16:00:49Z",
+       |"html_url":"http://gitbucket.exmple.com/octocat/Hello-World/issues/1347#comment-1"
+       |}""".stripMargin
+
+  val jsonCommentPR = s"""{
+       |"id":1,
+       |"user":$jsonUser,
+       |"body":"Me too",
+       |"created_at":"2011-04-14T16:00:49Z",
+       |"updated_at":"2011-04-14T16:00:49Z",
+       |"html_url":"http://gitbucket.exmple.com/octocat/Hello-World/pull/1347#comment-1"
+       |}""".stripMargin
+
+  val jsonCommitListItem = s"""{
+       |"sha":"6dcb09b5b57875f334f61aebed695e2e4193db5e",
+       |"commit":{
+         |"message":"full message",
+         |"author":{"name":"octocat","email":"octocat@example.com","date":"2011-04-14T16:00:49Z"},
+         |"committer":{"name":"octocat","email":"octocat@example.com","date":"2011-04-14T16:00:49Z"},
+         |"url":"http://gitbucket.exmple.com/api/v3/repos/octocat/Hello-World/git/commits/6dcb09b5b57875f334f61aebed695e2e4193db5e"
+       |},
+       |"parents":[{
+         |"sha":"1da452aa92d7db1bc093d266c80a69857718c406",
+         |"url":"http://gitbucket.exmple.com/api/v3/repos/octocat/Hello-World/commits/1da452aa92d7db1bc093d266c80a69857718c406"}],
+       |"url":"http://gitbucket.exmple.com/api/v3/repos/octocat/Hello-World/commits/6dcb09b5b57875f334f61aebed695e2e4193db5e"
+       |}""".stripMargin
+
+  // TODO url is correct?
+  val jsonCommit = (id: String) => s"""{
+       |"id":"$id",
+       |"message":"full message",
+       |"timestamp":"2011-04-14T16:00:49Z",
+       |"added":[],
+       |"removed":[],
+       |"modified":["README.md"],
+       |"author":{"name":"octocat","email":"octocat@example.com","date":"2011-04-14T16:00:49Z"},
+       |"committer":{"name":"octocat","email":"octocat@example.com","date":"2011-04-14T16:00:49Z"},
+       |"url":"http://gitbucket.exmple.com/octocat/Hello-World/commit/$id"
```

Comment:
`url` doesn't include path `api/v3/repos`. Is this correct?

Reply:
It's correct in `WebHookJsonFormatSpec`, but not correct in `JsonFormatSpec`.

In this case, `apiCommit` is created as follows:
```scala
  val apiCommit = {
    val commit = commitInfo(sha1)
    ApiCommit(
      id = commit.id,
      message = commit.fullMessage,
      timestamp = commit.commitTime,
      added = Nil,
      removed = Nil,
      modified = List("README.md"),
      author = ApiPersonIdent.author(commit),
      committer = ApiPersonIdent.committer(commit)
    )(repo1Name, true)
  }
```
The last parameter (true) is `urlIsHtmlUrl`. This parameter should be `true` when this model is used as web hook payload, and should be `false` when this mode is used as response of API.

So, ideally we might have to define two models. One is for `JsonFormatSpec` created with `urlIsHtmlUrl = false` and another one is for `WebHookJsonFormatSpec` created with `urlIsHtmlUrl = true`.

Reply:
https://github.com/gitbucket/gitbucket/blob/master/src/main/scala/gitbucket/core/api/ApiCommit.scala#L26-L35

Only this switches output attributes, and the others output both `url` and `html_url`. GitHub API also looks like it outputs both.